### ISSUE TITLE
drush_user_abort() does not set an error exit code

### ISSUE THREAD

**Issue by swichers** [drush-ops/drush#1265]
If a drush command, for example `sql-drop`, is aborted by a user drush will return a success exit code. This makes scripting interactive solutions very difficult, as stderr must be captured and parsed.

The specific use case I am running in to is that I am scripting some drush commands as part of a build process. There are some specific commands that I would like the user to give confirmation to before proceeding (hence no yes-to-all). Drush currently provides no easy way (that I can find) for me to know a user abort vs regular execution path completed without scraping text.

I can understand the logic behind having a user abort treated as a success (the script did not fail). I think an `--abort-is-failure` argument would be very useful. If there is already some hidden command that provides this behavior feel free to let me know.

**Comment by greg-1-anderson** [comment]
Your use case seems reasonable; however, we have this precidence:
```
$ rm -i foo
Remove foo? n
$ echo $?
0
```
It is sort of my inclination to buck the trend, and make drush_user_abort() set a specific error code. It seems that in the typical case, you would want the script to exit if the user cancelled a command.

Thoughts?

**Comment by swichers** [comment]
I can say I can't think of a decently good use case for continuing a script after a user aborts part of it. I suppose it is possible someone would want to skip rsyncing the files directory, but continue with the rest of a script.

That said, I'm in favor of maintaining consistency with other shell utilities. I think the best compromise is an option that can be used when calling drush that will make it consider aborts as failures.

**Comment by weitzman** [comment]
If you are scripting in php, it is trivial (once you know the right bits) to recognize an abort. 

    $return = drush_invoke_process('@self', 'sql-drop', ...);

Now, `$return['context']` is an array which lists the abort clearly. If you are not seeing that, please post here. 

I'm not convinced that further action is needed here.

**Comment by swichers** [comment]
I should clarify that by scripting, I do not mean a php script that is being executed by drush or executing internal drush commands. I mean a script (or program) that is executing the drush process directly. Even just chaining bash commands.

    drush sql-drop && echo 'Success!'

If you answer no to the prompt, you will see 'Success!' because technically the script executed without error even though the action you wanted your script to take was not completed. If we had something like

    drush --abort-is-failure sql-drop && echo 'Success!'

Then we shouldn't see 'Success!' when the user aborted, and our script could handle the fail condition accordingly.

Does that clear things up?

**Comment by weitzman** [comment]
Thanks for clarifying your use case. Scripting, combined with user interaction, is not a case we encounter frequently.

I'm a bit opposed to --abort-on-failure since there are so many global options that folks want to see for their use case. It gets awkward to list them all and harder to support them all. 

I'm not sure how disruptive it would to always error on abort. It sounds OK to me.

**Comment by greg-1-anderson** [comment]
I also don't like the the global option, for the same reason.  If we set the error status on user abort, it would still be possible for scripts to trap the error and keep running, if they wanted to do that.  I think that continuing after the user abort is more of the edge case, and I'm not sure why existing tools don't set an error code on user aborts.

There are two things we might do here.

1. For php users, it seems that it might be helpful if we set something more obvious than $return['context'] to signal user aborts.  Maybe $return['aborted']?

2. I am somewhat undecided about setting an error status on abort.  Sort of leaning towards thinking its a good idea.
